solidity/contracts/zeto_anon_nullifier_kyc.sol

```diff
@@ -0,0 +1,138 @@
+        ZetoFungibleWithdrawWithNullifiers(_depositVerifier, _withdrawVerifier)
+    {
+        verifier = _verifier;
+    }
+
+    function register(uint256[2] memory publicKey) public onlyOwner {
+        _register(publicKey);
+    }
+
+    /**
+     * @dev the main function of the contract.
+     *
+     * @param nullifiers Array of nullifiers that are secretly bound to UTXOs to be spent by the transaction.
+     * @param outputs Array of new UTXOs to generate, for future transactions to spend.
+     * @param root The root hash of the Sparse Merkle Tree that contains the nullifiers.
+     * @param proof A zero knowledge proof that the submitter is authorized to spend the inputs, and
+     *      that the outputs are valid in terms of obeying mass conservation rules.
+     *
+     * Emits a {UTXOTransfer} event.
+     */
+    function transfer(
+        uint256[2] memory nullifiers,
+        uint256[2] memory outputs,
+        uint256 root,
+        Commonlib.Proof calldata proof
+    ) public returns (bool) {
+        require(
+            validateTransactionProposal(nullifiers, outputs, root),
+            "Invalid transaction proposal"
+        );
+
+        // construct the public inputs
+        uint256[8] memory publicInputs;
+        publicInputs[0] = nullifiers[0];
+        publicInputs[1] = nullifiers[1];
+        publicInputs[2] = root;
+        publicInputs[3] = (nullifiers[0] == 0) ? 0 : 1; // if the first nullifier is empty, disable its MT proof verification
+        publicInputs[4] = (nullifiers[1] == 0) ? 0 : 1; // if the second nullifier is empty, disable its MT proof verification
+        publicInputs[5] = getIdentitiesRoot();
+        publicInputs[6] = outputs[0];
+        publicInputs[7] = outputs[1];
+
+        // // Check the proof
+        require(
+            verifier.verifyProof(proof.pA, proof.pB, proof.pC, publicInputs),
+            "Invalid proof"
+        );
+
+        processInputsAndOutputs(nullifiers, outputs);
+
+        uint256[] memory nullifierArray = new uint256[](nullifiers.length);
+        uint256[] memory outputArray = new uint256[](outputs.length);
+        for (uint256 i = 0; i < nullifiers.length; ++i) {
+            nullifierArray[i] = nullifiers[i];
+            outputArray[i] = outputs[i];
+        }
+        emit UTXOTransfer(nullifierArray, outputArray, msg.sender);
+        return true;
+    }
+
+    function deposit(
+        uint256 amount,
+        uint256 utxo,
+        Commonlib.Proof calldata proof
+    ) public {
+        _deposit(amount, utxo, proof);
+        uint256[] memory utxos = new uint256[](1);
+        utxos[0] = utxo;
+        _mint(utxos);
+    }
+
+    function withdraw(
+        uint256 amount,
+        uint256[2] memory nullifiers,
+        uint256 output,
+        uint256 root,
+        Commonlib.Proof calldata proof
+    ) public {
+        _withdrawWithNullifiers(amount, nullifiers, output, root, proof);
+        processInputsAndOutputs(nullifiers, [output, 0]);
+    }
```

Comment:
Should these 2 functions be allowed if the sender is not in the registry?

Reply:
correct, even if an unregistered user did a deposit, the only thing that user can do is return it back out via withdraw because the UTXOs are not allowed to be transferred.